vscode/src/services/LocalStorageProvider.ts

```
@@ -67,34 +71,80 @@ export class LocalStorage {
         await this.storage.update(this.CODY_ENDPOINT_HISTORY, [...historySet])
     }
 
-    public getChatHistory(): UserLocalHistory {
-        const history = this.storage.get<UserLocalHistory | null>(this.KEY_LOCAL_HISTORY, null)
-        return history || { chat: {}, input: [] }
+    public getChatHistory(authStatus: AuthStatus): UserLocalHistory {
+        let history = this.storage.get<{ [key: ChatHistoryKey]: UserLocalHistory } | UserLocalHistory | null>(
+            this.KEY_LOCAL_HISTORY,
+            null
+        )
+        if (!history) {
+            return { chat: {}, input: [] }
+        }
+
+        const key = getKeyForAuthStatus(authStatus)
+
+        // For backwards compatibility, we upgrade the local storage key from the old layout that is
+        // not scoped to individual user accounts to be scoped instead.
+        if (history && !isChatHistoryV2(history)) {
+            history = {
+                [key]: history as UserLocalHistory,
+
+                // For backward compatibility, we do not want to delete the `chat` and `input` keys.
+                // As otherwise, downgrading to a prior version would completely block the startup
+                // as the client would throw.
+                chat: {},
+                input: [],
+            }
+
+            // We use a raw write here to ensure we do not _append_ a key but actually replace
+            // existing `chat` and `input` keys.
+            void this.storage.update(this.KEY_LOCAL_HISTORY, history)
```

Do we want to `try/catch` or `await` here, similar to how it's done in `setChatHistory`?

---

That function is non-blocking right now and I wanted to avoid changing the API (since it is called in some constructors AFAIK). We definitely don't want to silence the error though but I think having it write asynchronously is fine. The worst part we call the same function again and the state is not updated yet but that would simply call the same update again and not destruct anything (provided writes are atomic but if they are not it would be v strange I guess?).